data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,675 @@
+- Functions returning a function, such as `sweet_cat` and `sour_cat`
+- Functions taking a tuple as a parameter, such as `spicy_cat`
+
+Interestingly, both kinds of functions provide a way to pass several pieces of data to a body while being functions with a single parameter. From this perspective, it makes sense to say: “All functions have a single argument.”
+
+This goes even further. It is always possible to translate back and forth between functions that look like `sweet_cat` (or `sour_cat`) and functions that look like `spicy_cat`.
+
+These translations have names:
+* [Currying](https://en.wikipedia.org/wiki/Currying) goes from the `spicy_cat` form into the `sour_cat` (or `sweet_cat`) form.
+* Uncurrying goes from the `sour_cat` (or `sweet_cat`) form into the `spicy_cat` form.
+
+It also said that `sweet_cat` and `sour_cat` are _curried_ functions whilst `spicy_cat` is _uncurried_.
+
+These translations are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry).
+
+From a typing perspective, this means the following types are equivalent:
+- `string -> (string -> string)` &mdash; curried function type
+- `string * string -> string` &mdash; uncurried function type
+
+Here, this is shown using `string` as an example, but it applies to any group of three types.
+
+We will not dive any deeper into the details here, but these equivalences can be formally defined using _ad-hoc_ mathematics.
+
+It is rarely necessary to use the functions above. However, it is important to understand that changing a function may not need any code refactoring. A function can be given several equivalent forms and changed into another, either using refactoring or using a higher-order function. Since functions are values, currying and uncurrying are operations on those values.
+
+In practice, curried functions are the default form functions should take because:
+- They allow partial application
+- Less editing, no parentheses or commas
+- No pattern matching over a tuple
+
+## Functions With Side-Effects
+
+With respect to its type, a function is processes input data from its *domain*, the set of input values, and produce a result data from its *codomain*, the set of output values.
+```ocaml
+# string_of_int;;
+- : int -> string = <fun>
+```
+The function `string_of_int`
+* Has domain `int`
+* Has codomain `string`
+
+However, some functions either take input data outside of their domain or produce data outside of their codomain. These out-of-signature data are called effects, or side effects. Input and output (I/O) are the most common forms of effects. Input is out-of-domain data and output is out-of-codomain data. However, the result of functions returning random numbers or the current time is influenced by external factors, which is also called an effect. The external factor is out-of-domain input. Similarly, any observable phenomena triggered by the computation of a function is out-of-codomain output.
```

Comment:
I find this section on domains/codomains and effects rather confusing. Perhaps we can review it in the meeting.

Regardless, "returning random numbers or the current time" really doesn't make sense to me.

Reply:
> returning random numbers (such as `Random.bits` does) or the current time (such as `Unix.time` does)